Detectors/ITSMFT/MFT/base/src/GeometryTGeo.cxx

```diff
@@ -333,148 +378,208 @@ TGeoHMatrix* GeometryTGeo::extractMatrixSensor(Int_t index) const
     gGeoManager->PopPath();
     LOG(ERROR) << "Error in cd-ing to " << path.Data() << FairLogger::endl;
     return nullptr;
-  }
+  } // end if !gGeoManager
 
   matTmp = *gGeoManager->GetCurrentMatrix(); // matrix may change after cd
 
   // Restore the modeler state.
   gGeoManager->PopPath();
 
-  return &matTmp;
+  Double_t *rot1 = mTransMFT2ITS->GetRotationMatrix();
+
+  Double_t rot2[9] = {0.,0.,0.,0.,0.,0.,0.,0.,0.};
+
+  TGeoHMatrix* matTmp1 = new TGeoHMatrix(matTmp);
+  Double_t *rot3 = matTmp1->GetRotationMatrix();
+
+  for (Int_t i = 0; i < 3; i++) {
+    for (Int_t j = 0; j < 3; j++) {
+      rot2[i*3+j] = 0.;
+      for (Int_t k = 0; k < 3; k++) {
+	rot2[i*3+j] += rot1[i*3+k]*rot3[k*3+j];
+      }
+    }
+  }
+  matTmp1->SetRotation(rot2);
+
+  return matTmp1;
 
 }
 
-//_____________________________________________________________________________
-void GeometryTGeo::build(Bool_t loadSegmentations)
+//__________________________________________________________________________
+void GeometryTGeo::fillMatrixCache(Int_t mask)
 {
 
-  if (!gGeoManager) {
-    LOG(FATAL) << "Geometry is not loaded" << FairLogger::endl;
-  } 
-  Geometry *mftGeo = Geometry::instance();
-
-  Segmentation *seg;
-  if (loadSegmentations) {
-    LOG(INFO) << "GeometryTGeo::build load segmentation from xml file!" << FairLogger::endl;
-    seg = new Segmentation(gSystem->ExpandPathName("$(VMCWORKDIR)/Detectors/Geometry/MFT/data/Geometry.xml" ));
-  } else {
-    LOG(INFO) << "GeometryTGeo::build get segmentation from the geometry!" << FairLogger::endl;
-    seg = mftGeo->getSegmentation();
+  // populate matrix cache for requested transformations
+  //  
+  if (mSize < 1) {    
+    LOG(WARNING) << "The method Build was not called yet" << FairLogger::endl;
+    Build(true,mask);
+    return;
   }
-
-  mNumberOfDisks = Constants::sNDisks;
-
-  // fill the map <sensorUniqueID,chipID>
-  Int_t chipID = 0;
-  Int_t sensorUniqueID;
-  for (Int_t iHalf = 0; iHalf < 2; iHalf++) {
-    HalfSegmentation * halfSeg = seg->getHalf(iHalf);
-    for (Int_t iDisk = 0; iDisk < mNumberOfDisks; iDisk++) {
-      HalfDiskSegmentation* halfDiskSeg = halfSeg->getHalfDisk(iDisk);
-      for (Int_t iLadder = 0; iLadder < halfDiskSeg->getNLadders(); iLadder++) {
-	LadderSegmentation* ladderSeg = halfDiskSeg->getLadder(iLadder);
-	mNumberOfChips += ladderSeg->getNSensors();
-	for (Int_t iSensor = 0; iSensor < ladderSeg->getNSensors(); iSensor++) {
-	  ChipSegmentation * chipSeg = ladderSeg->getSensor(iSensor);
-	  sensorUniqueID = chipSeg->GetUniqueID();
-	  mChipIDmap.insert(BimapType::value_type(sensorUniqueID,chipID++));
-	}
-      }
+  //LOG(INFO) << "mask " << mask << " bit2Mask " << bit2Mask(TransformType::L2G) << FairLogger::endl;
+  // build matrices
+  if ( (mask & bit2Mask(TransformType::L2G)) && !getCacheL2G().isFilled() ) {
+    LOG(INFO) << "Loading ITS L2G matrices from TGeo" <<  FairLogger::endl;
```

Comment:
Better to use MFT instead of ITS in the logs (there are more instances like that).